# **Mandriva Website**
Mandriva is a genomic analysis web application specializing in the analysis of Single Nucleotide Polymorphism (SNP) variations and population genetics specifically on human chromosome 1. The platform encompasses a range of analytical features including clustering analysis, admixture analysis, and FST analysis, complemented by visualizations. Additionally, Mandriva delivers insights into genotype and allele frequenciesa, as well as the clinical relevance associated with selected SNPs.


## Installing Mandriva
### Data
* Please download the required database (mandriva_database.db) from [Mandriva](https://qmulprod-my.sharepoint.com/:f:/g/personal/bt23629_qmul_ac_uk/EuwCtRYgQb9PlN-T18AO-54B21IENhkSLMXAFRdZJxm9xQ?e=h9aLxJ) (QMUL email is required to access the data)
* If you want to visualize the database, a DB Browser for SQLite is required, which can be downloaded from [here](https://sqlitebrowser.org/dl/) 


### Instructions 
1. `git clone https://github.com/TomBennett2202/Mandriva.git`
2. `cd Mandriva`
3. `pip3 install -r requirements.txt`
(QMUL email is required to access the data)
* If you want to visualize the database, a DB Browser for SQLite is required, which can be downloaded from [here](https://sqlitebrowser.org/dl/) 


### Instructions 

1. `git clone https://github.com/TomBennett2202/Mandriva.git`
2. `cd Mandriva`
3. `pip3 install -r requirements.txt`


* Please ensure that the mandriva_database.db and app.py are in the same folder while running the app.
If you wish to change the database directory, you should also change the *DATABASE* path in the application.



## Developers
This web application was designed and developed by [Thomas Bennett](https://github.com/TomBennett2202), [Melika Abdi](https://github.com/meliabdi), [Niya Louis](https://github.com/Niyalouis) & [Adam Brachtl](https://github.com/ABrachtl) as part of the 2023/24 MSc Bioinformatics Software Development Group Project at Queen Mary University of London. 

Supervisors: Professor Conrad Bessant & Dr Matteo Fumagalli.



